main.py

```diff
@@ -1,23 +1,137 @@
+
+    if request.method == "POST":
+        if float(request.form["value"]) < 0 or float(request.form["value"]) > 10**10:
+            return render_template(
+                "add.jinja2", error="Please enter a reasonable number..."
+            )
+        try:
+            existing_donor = Donor.get(Donor.name == request.form["donor"].title())
+            Donation(donor=existing_donor, value=request.form["value"]).save()
+        except DoesNotExist:
+            new_donor = Donor(name=request.form["donor"].title())
+            new_donor.save()
+            Donation(donor=new_donor, value=request.form["value"]).save()
+        return redirect(url_for("all_donations"))
+
+    return render_template("add.jinja2")
+
+
+@app.route("/login", methods=["GET", "POST"])
+def login():
+    """ Login page for users, required to add donation """
+    if request.method == "POST":
+        try:
+            user = User.select().where(User.name == request.form["name"]).get()
+            if user and pbkdf2_sha256.verify(request.form["password"], user.password):
+                session["logged_in"] = True
+                return redirect(url_for("all_donations"))
+            return render_template(
+                "login.jinja2", error="Incorrect username or password."
+            )
+        except DoesNotExist:
+            return render_template(
+                "login.jinja2", error="Incorrect username or password."
+            )
+    return render_template("login.jinja2")
+
+
+@app.route("/logout")
+def logout():
+    """ Logout page.
+
+    Sets session['logged_in'] = False and returns all_donation page
+    """
+    session["logged_in"] = False
+    return redirect(url_for("login"))
+
+
+@app.route("/donors", methods=["GET", "POST"])
+def donors():
+    """ Lists donors, and donations for selected donor """
+    if not logged_in():
+        return redirect(url_for("login"))
+
+    if request.method == "POST":
+        try:
+            return render_template(
+                "donors.jinja2",
+                donors=Donor.select(),
+                donations=Donation.select().where(
+                    Donation.donor == Donor.get(
+                        Donor.name == request.form["donor"].title()
+                    )
+                ),
+                name=request.form["donor"].title(),
+            )
+        except DoesNotExist:
+            return render_template(
+                "donors.jinja2",
+                donors=Donor.select(),
+                error=f"Donor {request.form['donor']} not found.",
+            )
+
+    return render_template("donors.jinja2", donors=Donor.select())
+
+
+def logged_in():
+    """ Determine if user is currently logged in.
+
+    Returns True if logged_in, False otherwise.
+    """
+    try:
+        if session["logged_in"]:
+            return True
+    except KeyError:
+        # Tried initializing session in main, but that doesn't work.
+        # How to default session["logged_in"] = False ???
```

Comment:
When we start to format our projects in a more typical flask directory structure, you would do it in the __init__.py file which would handle the stuff in your `main()` function.  Also in your login piece, you would create a session['profile'] which would hold all the information, including checking for a cookie token to associate with the profile so you know if the user is authenticated or not.  The lack of that cookie would cause you to prompt them to login.